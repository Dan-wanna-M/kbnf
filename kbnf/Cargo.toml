[package]
name = "kbnf"
version = "0.1.0"
edition = "2021"

[dependencies]
ahash = "0.8.11"
thiserror = "1.0.61"
ebnf = {git="https://github.com/Dan-wanna-M/ebnf.git"}
num = "0.4.3"
jaggedarray = "0.2.2"
string-interner = "0.17.0"
nom = "7.1.3"
tinyvec = "1.6.0"
nonmax = "0.5.5"
fixedbitset = {git="https://github.com/Dan-wanna-M/fixedbitset.git"}
regex-automata={git="https://github.com/Dan-wanna-M/regex.git"}
displaydoc = "0.2.4"
serde = "1.0.203"
strum = { version = "0.26", features = ["derive"] }
[dev-dependencies]
insta = { version = "1.26.0"}
serde_json = "1.0.48"