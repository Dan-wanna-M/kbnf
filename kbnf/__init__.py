from . import *
from engine import *