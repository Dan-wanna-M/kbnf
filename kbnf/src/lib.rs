pub mod engine;
pub mod utils;
pub mod vocabulary;
mod generic_rc;
pub mod grammar;
pub mod config;

pub fn add(left: usize, right: usize) -> usize {
    left + right
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let result = add(2, 2);
        assert_eq!(result, 4);
    }
}
