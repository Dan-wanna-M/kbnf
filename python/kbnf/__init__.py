from .engine import Engine
from .kbnf import *


__doc__ = kbnf.__doc__
if hasattr(kbnf, "__all__"):
    __all__ = kbnf.__all__