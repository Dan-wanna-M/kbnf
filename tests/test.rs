#[cfg(test)]

mod tests {
    use std::{
        cell::RefCell,
        fs::File,
        io::BufReader,
        path::Path,
        sync::{Arc, Mutex},
    };

    use ahash::AHashMap;
    use insta::assert_snapshot;
    use kbnf::{
        engine::EngineConfig,
        engine_like::{AcceptTokenResult, EngineLike},
        vocabulary::{Token, Vocabulary},
    };
    #[derive(Debug, thiserror::Error)]
    /// Error type when reading RWKV world model's vocabulary file.
    pub enum ReadRWKVVocabError {
        #[error("IO error: {0}")]
        /// Error due to I/O operations like [Read], [Write], [Seek],
        IoError(#[from] std::io::Error),
        #[error("Serde json error: {0}")]
        JsonError(#[from] serde_json::Error),
    }

    /// Read the vocabulary from RWKV-world model series vocabulary file.
    pub fn read_rwkv_world_vocab(path: impl AsRef<Path>) -> Result<Vocabulary, ReadRWKVVocabError> {
        let path = path.as_ref();
        let file = File::open(path).unwrap();
        let reader = BufReader::new(file);
        let mut id_to_token: AHashMap<u32, Token> = AHashMap::default();
        let mut id_to_token_string: AHashMap<u32, String> = AHashMap::default();
        let data: serde_json::Map<String, serde_json::Value> = serde_json::from_reader(reader)?;
        for (key, value) in data {
            let key = key.parse::<u32>().unwrap();
            match value {
                serde_json::Value::Array(x) => {
                    let mut token = Vec::new();
                    for x in x {
                        match x {
                            serde_json::Value::Number(x) => {
                                token.push(x.as_u64().unwrap() as u8);
                            }
                            _ => {
                                panic!("Unexpected value type")
                            }
                        }
                    }
                    id_to_token.insert(key, Token(token.clone().into_boxed_slice()));
                    id_to_token_string.insert(key, format!("{:?}", token));
                }
                serde_json::Value::String(x) => {
                    id_to_token.insert(key, Token(x.as_bytes().to_vec().into_boxed_slice()));
                    id_to_token_string.insert(key, x);
                }
                _ => {
                    panic!("Unexpected value type")
                }
            };
        }
        Ok(Vocabulary::new(id_to_token, id_to_token_string).unwrap())
    }

    fn get_token_id_from_str(vocab: &Vocabulary, token: &str) -> Option<u32> {
        vocab.token_id(&Token(token.as_bytes().to_vec().into_boxed_slice()))
    }
    #[test]
    fn single_terminal() {
        let input = "start::='Hello, World!\n';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        assert!(
            engine.try_accept_new_token(get_token_id_from_str(&vocab, "b").unwrap())
                == Err(kbnf::engine_like::AcceptTokenError::Rejected),
            "This should not be accepted"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, "Hello").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, ",").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        assert!(
            !engine.allowed_token_ids_from_last_computation().is_empty(),
            "allowed token ids are not updated correctly!"
        );
    }

    #[test]
    fn single_regex() {
        let input = "start::=#'Hello, World!\n';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        assert!(
            engine.try_accept_new_token(get_token_id_from_str(&vocab, "b").unwrap())
                == Err(kbnf::engine_like::AcceptTokenError::Rejected),
            "This should not be accepted"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, "Hello").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, ",").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        assert!(
            !engine.allowed_token_ids_from_last_computation().is_empty(),
            "allowed token ids are not updated correctly!"
        );
    }
    #[test]
    fn single_regex2() {
        let input = "start::=#'[0-9]+''\\n';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let config = kbnf::config::Config {
            engine_config: EngineConfig {
                cache_enabled: true,
                compaction_enabled: false,
            },
            ..Default::default()
        };
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::with_config(input, vocab.clone(), config).unwrap();
        assert!(
            engine.try_accept_new_token(get_token_id_from_str(&vocab, "b").unwrap())
                == Err(kbnf::engine_like::AcceptTokenError::Rejected),
            "This should not be accepted"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, "1").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine.try_accept_new_token(get_token_id_from_str(&vocab, ",").unwrap());
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            result == Err(kbnf::engine_like::AcceptTokenError::Rejected),
            "Token should not be accepted"
        );
        assert!(
            !engine.allowed_token_ids_from_last_computation().is_empty(),
            "allowed token ids are not updated correctly!"
        );
    }

    #[test]
    fn minimal_case() {
        let input = "start::='aaa';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        assert!(
            engine.try_accept_new_token(get_token_id_from_str(&vocab, "b").unwrap())
                == Err(kbnf::engine_like::AcceptTokenError::Rejected),
            "This should not be accepted"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, "a").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, "a").unwrap())
                .unwrap()
                == AcceptTokenResult::Ongoing,
            "Failed to accept token"
        );
        engine.compute_allowed_token_ids();
        assert!(
            engine
                .try_accept_new_token(get_token_id_from_str(&vocab, "a").unwrap())
                .unwrap()
                == AcceptTokenResult::Finished,
            "Failed to accept token"
        );
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
    }

    #[test]
    fn escaped_literal() {
        let input = "start::=#'(\\n\\n)+';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        for i in 0..10 {
            engine.compute_allowed_token_ids();
            assert!(
                !engine.allowed_token_ids_from_last_computation().is_empty(),
                "Allowed token ids are not updated correctly!"
            );
            assert!(
                engine.try_accept_new_token(get_token_id_from_str(&vocab, "b").unwrap())
                    == Err(kbnf::engine_like::AcceptTokenError::Rejected),
                "This should not be accepted"
            );
            engine.compute_allowed_token_ids();
            assert!(
                engine
                    .try_accept_new_token(get_token_id_from_str(&vocab, "\n\n").unwrap())
                    .unwrap()
                    == AcceptTokenResult::Finished,
                "Failed to accept token"
            );
            engine.compute_allowed_token_ids();
            engine.reset();
        }
    }

    #[test]
    fn left_recursion() {
        let input = "start::='bb'|start'bb';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("bb".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_snapshot!(format!("{:#?}", engine));
        assert_eq!(result, AcceptTokenResult::Finished);
    }

    #[test]
    fn right_recursion() {
        let input = "start::=C'\n';C::='c'|#'c' C;";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let config = kbnf::config::Config {
            engine_config: EngineConfig {
                cache_enabled: true,
                compaction_enabled: true,
            },
            ..Default::default()
        };
        let mut engine = kbnf::engine::Engine::with_config(input, vocab.clone(), config).unwrap();
        for i in 0..10 {
            let result = engine
                .try_accept_new_token(
                    vocab
                        .token_id(&Token("c".as_bytes().to_vec().into_boxed_slice()))
                        .unwrap(),
                )
                .unwrap();
            assert_eq!(result, AcceptTokenResult::Ongoing);
            // engine.compute_allowed_token_ids();
        }
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("\n".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Finished);
        assert_snapshot!(format!("{:#?}", engine));
    }

    #[test]
    fn escaped_character() {
        let input = "start::=C'\n';C::='\\u0020'| #'\\u0020' C;";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let config = kbnf::config::Config {
            engine_config: EngineConfig {
                cache_enabled: true,
                compaction_enabled: true,
            },
            ..Default::default()
        };
        let mut engine = kbnf::engine::Engine::with_config(input, vocab.clone(), config).unwrap();
        for i in 0..10 {
            let result = engine
                .try_accept_new_token(
                    vocab
                        .token_id(&Token("\u{0020}".as_bytes().to_vec().into_boxed_slice()))
                        .unwrap(),
                )
                .unwrap();
            assert_eq!(result, AcceptTokenResult::Ongoing);
            // engine.compute_allowed_token_ids();
        }
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("\n".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Finished);
        assert_snapshot!(format!("{:#?}", engine));
    }
    #[test]
    fn indirect_right_recursion() {
        let input = "start::=A'\n';A::='x'|'x' B;B::='y'|'y' A;";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let config = kbnf::config::Config {
            engine_config: EngineConfig {
                cache_enabled: true,
                compaction_enabled: true,
            },
            ..Default::default()
        };
        let mut engine = kbnf::engine::Engine::with_config(input, vocab.clone(), config).unwrap();
        for i in 0..10 {
            let value = if i % 2 == 0 { "x" } else { "y" };
            let result = engine
                .try_accept_new_token(
                    vocab
                        .token_id(&Token(value.as_bytes().to_vec().into_boxed_slice()))
                        .unwrap(),
                )
                .unwrap();
            assert_eq!(result, AcceptTokenResult::Ongoing);
            // engine.compute_allowed_token_ids();
        }
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("\n".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Finished);
        assert_snapshot!(format!("{:#?}", engine));
    }
    #[test]
    fn middle_recursion() {
        let input = "start::=('{'start'}')?;";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        for _ in 0..10 {
            let result = engine
                .try_accept_new_token(
                    vocab
                        .token_id(&Token("{".as_bytes().to_vec().into_boxed_slice()))
                        .unwrap(),
                )
                .unwrap();
            assert_eq!(result, AcceptTokenResult::Ongoing);
            engine.compute_allowed_token_ids();
        }
        for _ in 0..9 {
            let result = engine
                .try_accept_new_token(
                    vocab
                        .token_id(&Token("}".as_bytes().to_vec().into_boxed_slice()))
                        .unwrap(),
                )
                .unwrap();
            assert_eq!(result, AcceptTokenResult::Ongoing);
            engine.compute_allowed_token_ids();
        }
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("}".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        // assert_snapshot!(format!("{:#?}", engine));
        assert_eq!(result, AcceptTokenResult::Finished);
    }
    #[test]
    fn always_match_regex() {
        let input = "start::=#\".+\"'\n';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        for j in 0..1 {
            for i in 0..5 {
                let result = engine
                    .try_accept_new_token(
                        vocab
                            .token_id(&Token("imper".as_bytes().to_vec().into_boxed_slice()))
                            .unwrap(),
                    )
                    .unwrap();
                assert_eq!(result, AcceptTokenResult::Ongoing);
                engine.compute_allowed_token_ids();
            }
            let result = engine
                .try_accept_new_token(
                    vocab
                        .token_id(&Token("\n".as_bytes().to_vec().into_boxed_slice()))
                        .unwrap(),
                )
                .unwrap();
            assert_eq!(result, AcceptTokenResult::Finished);
            engine.reset();
        }
    }

    #[test]
    fn substrings() {
        let input = "start::=#substrs'abcbc''\n';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("b".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Ongoing);
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("c".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Ongoing);
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("b".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Ongoing);
        let result = engine.try_accept_new_bytes(b"c").unwrap();
        let result = engine.try_accept_new_token(
            vocab
                .token_id(&Token("c".as_bytes().to_vec().into_boxed_slice()))
                .unwrap(),
        );
        assert_eq!(result, Err(kbnf::engine_like::AcceptTokenError::Rejected));
        engine.compute_allowed_token_ids();
        assert_snapshot!(format!("{:#?}", engine));
    }
    #[test]
    fn early_regex() {
        let input = "start::=#e'(.|\n)+\n\n''a';";
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(input, vocab.clone()).unwrap();
        for j in 0..1 {
            for i in 0..5 {
                let result = engine
                    .try_accept_new_token(
                        vocab
                            .token_id(&Token("imper".as_bytes().to_vec().into_boxed_slice()))
                            .unwrap(),
                    )
                    .unwrap();
                assert_eq!(result, AcceptTokenResult::Ongoing);
                engine.compute_allowed_token_ids();
            }
        }
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("\n".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Ongoing);
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("imper".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Ongoing);
        engine.compute_allowed_token_ids();
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("\n\n".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Ongoing);
        assert_eq!(
            engine.try_accept_new_token(
                vocab
                    .token_id(&Token("\n".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap()
            ),
            Err(kbnf::engine_like::AcceptTokenError::Rejected)
        );
        let result = engine
            .try_accept_new_token(
                vocab
                    .token_id(&Token("a".as_bytes().to_vec().into_boxed_slice()))
                    .unwrap(),
            )
            .unwrap();
        assert_eq!(result, AcceptTokenResult::Finished);
        engine.compute_allowed_token_ids();
        engine.reset();
    }

    #[test]
    fn linked_list() {
        let grammar_str = r#"__schema_json_1_next_0 ::= __schema_json_1;

start ::= "```json\n"__schema_json_1"```\n";

__schema_json_1 ::= 
    #"\\A\\{( |\n|\r|\t)*\\z" 
    "\"value\""
    #"\\A( |\n|\r|\t)*:( |\n|\r|\t)*\\z" 
    #"\\A-?(0|[1-9]\\d*)\\z" 
    #"\\A( |\n|\r|\t)*,( |\n|\r|\t)*\\z" 
    "\"next\"" 
    #"\\A( |\n|\r|\t)*:( |\n|\r|\t)*\\z" 
    __schema_json_1_next
    #"\\A( |\n|\r|\t)*\\}\\z";

__schema_json_1_next ::= 
    "null"
    | __schema_json_1_next_0;
"#;
        let vocab = read_rwkv_world_vocab("tests/rwkv_vocab_v20230424.json").unwrap();
        let logits = vec![0.0; vocab.vocab_size()];
        let mut engine = kbnf::engine::Engine::new(grammar_str, vocab.clone()).unwrap();
        engine
            .try_accept_new_bytes("```json\n{\"value\": 2, \"next\":".as_bytes())
            .unwrap();
        engine
            .try_accept_new_bytes(" {\"value\": 3, \"next\":null}".as_bytes())
            .unwrap();
    }
}
