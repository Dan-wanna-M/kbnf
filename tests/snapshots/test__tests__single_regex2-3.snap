---
source: tests/test.rs
expression: "format!(\"{:#?}\", engine)"
---
Engine {
    union: U16U16U32U32U32(
        EngineBase {
            grammar: Grammar {
                start_nonterminal: "start[0]",
                rules: "start[0] ::= #\"\\A(?:[0-9]+)\\z\"[0]\"\n\"[0];\n",
                id_to_regexes: {
                    "#\"\\A(?:[0-9]+)\\z\"[0]": Dfa(
                        dense::DFA(
                        D 000000: 
                        Q 000004: 
                         *000008: 
                          000012: \x01 => 20
                          000016: 
                          000020: \x01 => 20, EOI => 8
                        
                        START-GROUP(unanchored)
                          NonWordByte => 000016
                          WordByte => 000016
                          Text => 000012
                          LineLF => 000016
                          LineCR => 000016
                          CustomLineTerminator => 000016
                        START-GROUP(anchored)
                          NonWordByte => 000016
                          WordByte => 000016
                          Text => 000012
                          LineLF => 000016
                          LineCR => 000016
                          CustomLineTerminator => 000016
                        state length: 6
                        pattern length: 1
                        flags: Flags { has_empty: false, is_utf8: true, is_always_start_anchored: true }
                        )
                        ,
                    ),
                },
                id_to_suffix_automata: {},
                id_to_suffix_automata_first_bytes: [],
                id_to_regex_first_bytes: [
                    (
                        "#\"\\A(?:[0-9]+)\\z\"[0]",
                        StateID(
                            12,
                        ),
                        [
                            48,
                            49,
                            50,
                            51,
                            52,
                            53,
                            54,
                            55,
                            56,
                            57,
                        ],
                    ),
                    (
                        "#\"\\A(?:[0-9]+)\\z\"[0]",
                        StateID(
                            20,
                        ),
                        [
                            48,
                            49,
                            50,
                            51,
                            52,
                            53,
                            54,
                            55,
                            56,
                            57,
                        ],
                    ),
                ],
                id_to_regex_complement_first_bytes: [],
                id_to_terminals: [
                    (
                        "\"\n\"[0]",
                        [
                            10,
                        ],
                    ),
                ],
            },
            allowed_first_bytes: [
                10,
                48,
                49,
                50,
                51,
                52,
                53,
                54,
                55,
                56,
                57,
            ],
            allowed_token_ids: [
                "\n[11]",
                "0[49]",
                "1[50]",
                "2[51]",
                "3[52]",
                "4[53]",
                "5[54]",
                "6[55]",
                "7[56]",
                "8[57]",
                "9[58]",
                "00[620]",
                "01[621]",
                "02[622]",
                "03[623]",
                "04[624]",
                "05[625]",
                "06[626]",
                "07[627]",
                "08[628]",
                "09[629]",
                "10[630]",
                "11[631]",
                "12[632]",
                "13[633]",
                "14[634]",
                "15[635]",
                "16[636]",
                "17[637]",
                "18[638]",
                "19[639]",
                "20[640]",
                "21[641]",
                "22[642]",
                "23[643]",
                "24[644]",
                "25[645]",
                "26[646]",
                "27[647]",
                "28[648]",
                "29[649]",
                "30[650]",
                "31[651]",
                "32[652]",
                "33[653]",
                "34[654]",
                "35[655]",
                "36[656]",
                "37[657]",
                "38[658]",
                "39[659]",
                "40[660]",
                "41[661]",
                "42[662]",
                "43[663]",
                "44[664]",
                "45[665]",
                "46[666]",
                "47[667]",
                "48[668]",
                "49[669]",
                "50[670]",
                "51[671]",
                "52[672]",
                "53[673]",
                "54[674]",
                "55[675]",
                "56[676]",
                "57[677]",
                "58[678]",
                "59[679]",
                "60[680]",
                "61[681]",
                "62[682]",
                "63[683]",
                "64[684]",
                "65[685]",
                "66[686]",
                "67[687]",
                "68[688]",
                "69[689]",
                "70[690]",
                "71[691]",
                "72[692]",
                "73[693]",
                "74[694]",
                "75[695]",
                "76[696]",
                "77[697]",
                "78[698]",
                "79[699]",
                "80[700]",
                "81[701]",
                "82[702]",
                "83[703]",
                "84[704]",
                "85[705]",
                "86[706]",
                "87[707]",
                "88[708]",
                "89[709]",
                "90[710]",
                "91[711]",
                "92[712]",
                "93[713]",
                "94[714]",
                "95[715]",
                "96[716]",
                "97[717]",
                "98[718]",
                "99[719]",
            ],
            earley_sets: [
                [
                    EarleyItemDebugStruct {
                        dotted_rule: "start[0] -> .#\"\\A(?:[0-9]+)\\z\"[0]\"\n\"[0]",
                        start_position: 0,
                        state: "[3(InProgress)]",
                    },
                ],
                [
                    EarleyItemDebugStruct {
                        dotted_rule: "start[0] -> #\"\\A(?:[0-9]+)\\z\"[0].\"\n\"[0]",
                        start_position: 0,
                        state: "[0]",
                    },
                    EarleyItemDebugStruct {
                        dotted_rule: "start[0] -> .#\"\\A(?:[0-9]+)\\z\"[0]\"\n\"[0]",
                        start_position: 0,
                        state: "[5(Accept)]",
                    },
                ],
            ],
            cache: [
                (
                    [
                        [
                            EarleyItemDebugStruct {
                                dotted_rule: "start[0] -> .#\"\\A(?:[0-9]+)\\z\"[0]\"\n\"[0]",
                                start_position: 0,
                                state: "[3(InProgress)]",
                            },
                        ],
                    ],
                    (
                        [
                            "0[49]",
                            "1[50]",
                            "2[51]",
                            "3[52]",
                            "4[53]",
                            "5[54]",
                            "6[55]",
                            "7[56]",
                            "8[57]",
                            "9[58]",
                            "00[620]",
                            "01[621]",
                            "02[622]",
                            "03[623]",
                            "04[624]",
                            "05[625]",
                            "06[626]",
                            "07[627]",
                            "08[628]",
                            "09[629]",
                            "10[630]",
                            "11[631]",
                            "12[632]",
                            "13[633]",
                            "14[634]",
                            "15[635]",
                            "16[636]",
                            "17[637]",
                            "18[638]",
                            "19[639]",
                            "20[640]",
                            "21[641]",
                            "22[642]",
                            "23[643]",
                            "24[644]",
                            "25[645]",
                            "26[646]",
                            "27[647]",
                            "28[648]",
                            "29[649]",
                            "30[650]",
                            "31[651]",
                            "32[652]",
                            "33[653]",
                            "34[654]",
                            "35[655]",
                            "36[656]",
                            "37[657]",
                            "38[658]",
                            "39[659]",
                            "40[660]",
                            "41[661]",
                            "42[662]",
                            "43[663]",
                            "44[664]",
                            "45[665]",
                            "46[666]",
                            "47[667]",
                            "48[668]",
                            "49[669]",
                            "50[670]",
                            "51[671]",
                            "52[672]",
                            "53[673]",
                            "54[674]",
                            "55[675]",
                            "56[676]",
                            "57[677]",
                            "58[678]",
                            "59[679]",
                            "60[680]",
                            "61[681]",
                            "62[682]",
                            "63[683]",
                            "64[684]",
                            "65[685]",
                            "66[686]",
                            "67[687]",
                            "68[688]",
                            "69[689]",
                            "70[690]",
                            "71[691]",
                            "72[692]",
                            "73[693]",
                            "74[694]",
                            "75[695]",
                            "76[696]",
                            "77[697]",
                            "78[698]",
                            "79[699]",
                            "80[700]",
                            "81[701]",
                            "82[702]",
                            "83[703]",
                            "84[704]",
                            "85[705]",
                            "86[706]",
                            "87[707]",
                            "88[708]",
                            "89[709]",
                            "90[710]",
                            "91[711]",
                            "92[712]",
                            "93[713]",
                            "94[714]",
                            "95[715]",
                            "96[716]",
                            "97[717]",
                            "98[718]",
                            "99[719]",
                        ],
                    ),
                ),
                (
                    [
                        [
                            EarleyItemDebugStruct {
                                dotted_rule: "start[0] -> .#\"\\A(?:[0-9]+)\\z\"[0]\"\n\"[0]",
                                start_position: 0,
                                state: "[3(InProgress)]",
                            },
                        ],
                        [
                            EarleyItemDebugStruct {
                                dotted_rule: "start[0] -> #\"\\A(?:[0-9]+)\\z\"[0].\"\n\"[0]",
                                start_position: 0,
                                state: "[0]",
                            },
                            EarleyItemDebugStruct {
                                dotted_rule: "start[0] -> .#\"\\A(?:[0-9]+)\\z\"[0]\"\n\"[0]",
                                start_position: 0,
                                state: "[5(Accept)]",
                            },
                        ],
                    ],
                    (
                        [
                            "\n[11]",
                            "0[49]",
                            "1[50]",
                            "2[51]",
                            "3[52]",
                            "4[53]",
                            "5[54]",
                            "6[55]",
                            "7[56]",
                            "8[57]",
                            "9[58]",
                            "00[620]",
                            "01[621]",
                            "02[622]",
                            "03[623]",
                            "04[624]",
                            "05[625]",
                            "06[626]",
                            "07[627]",
                            "08[628]",
                            "09[629]",
                            "10[630]",
                            "11[631]",
                            "12[632]",
                            "13[633]",
                            "14[634]",
                            "15[635]",
                            "16[636]",
                            "17[637]",
                            "18[638]",
                            "19[639]",
                            "20[640]",
                            "21[641]",
                            "22[642]",
                            "23[643]",
                            "24[644]",
                            "25[645]",
                            "26[646]",
                            "27[647]",
                            "28[648]",
                            "29[649]",
                            "30[650]",
                            "31[651]",
                            "32[652]",
                            "33[653]",
                            "34[654]",
                            "35[655]",
                            "36[656]",
                            "37[657]",
                            "38[658]",
                            "39[659]",
                            "40[660]",
                            "41[661]",
                            "42[662]",
                            "43[663]",
                            "44[664]",
                            "45[665]",
                            "46[666]",
                            "47[667]",
                            "48[668]",
                            "49[669]",
                            "50[670]",
                            "51[671]",
                            "52[672]",
                            "53[673]",
                            "54[674]",
                            "55[675]",
                            "56[676]",
                            "57[677]",
                            "58[678]",
                            "59[679]",
                            "60[680]",
                            "61[681]",
                            "62[682]",
                            "63[683]",
                            "64[684]",
                            "65[685]",
                            "66[686]",
                            "67[687]",
                            "68[688]",
                            "69[689]",
                            "70[690]",
                            "71[691]",
                            "72[692]",
                            "73[693]",
                            "74[694]",
                            "75[695]",
                            "76[696]",
                            "77[697]",
                            "78[698]",
                            "79[699]",
                            "80[700]",
                            "81[701]",
                            "82[702]",
                            "83[703]",
                            "84[704]",
                            "85[705]",
                            "86[706]",
                            "87[707]",
                            "88[708]",
                            "89[709]",
                            "90[710]",
                            "91[711]",
                            "92[712]",
                            "93[713]",
                            "94[714]",
                            "95[715]",
                            "96[716]",
                            "97[717]",
                            "98[718]",
                            "99[719]",
                        ],
                    ),
                ),
            ],
            to_be_completed_items: [],
            to_be_completed_items_buffer: [],
            deduplication_buffer: [],
            postdot_items: [],
            column_to_postdot_items: [],
            postdot_items_since_last_commit: [],
            leo_items: [],
            leo_items_buffer: [],
            already_predicted_nonterminals: [],
            finished: false,
            config: EngineConfig {
                cache_enabled: true,
                compaction_enabled: false,
            },
        },
    ),
}
